File: java/leetcode/src/main/java/com/leetcode/graph/NumberOfIslands_200.java
package com.leetcode.graph;

public class NumberOfIslands_200 {
    private int[][] dirs = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    private int m = 0;
    private int n = 0;
    private boolean[][] visited = null;

    public int numIslands(char[][] grid) {
        m = grid.length;
        n = grid[0].length;
        visited= new boolean[m][n];

        int cnt = 0;

        for(int i = 0; i < m; i++) {
            for(int j = 0;j < n; j++) {
                if(grid[i][j] == '1' && !visited[i][j]) {
                    dfs(i, j, grid, visited);
                    cnt++;
                }
            }
        }

        return cnt;

    }

    private void dfs(int i, int j, char[][] grid, boolean[][] visited) {
        if (i < 0 || i >= m || j <0 || j >= n || grid[i][j] == '0' || visited[i][j]) return;
        visited[i][j] = true;
        for(int[] d : dirs)
            dfs(i + d[0], j + d[1], grid, visited);
    }
}

File: js/algrithm/huahua130/Graph/Number_of_Islands_200.js
/**
 * @param {character[][]} grid
 * @return {number}
 */
var numIslands = function(grid) {
    const dirs = [[1, 0], [-1, 0], [0, 1], [0, -1]]
    const m = grid.length
    const n = grid[0].length
    const visited = []
    for (let i = 0; i < m; i++) {
        visited.push(Array(n).fill(false))
    }

    const dfs = function(grid, i, j, visited) {
        visited[i][j] = true
        for (let dir of dirs) {
            let nr = dir[0] + i
            let nc = dir[1] + j
            if (nr >=0 && nr < m && nc >=0 && nc < n && !visited[nr][nc] && grid[nr][nc] === "1") {
                dfs(grid, nr, nc, visited)
            }
        }
    }


    let cnt = 0
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (!visited[i][j] && grid[i][j] === "1") {
                dfs(grid, i, j, visited)
                cnt++
            }
        }
    }

    return cnt
};

